File: sdks/ruby/lib/openrate/ffi.rb
# frozen_string_literal: true

# openrate in-process, through the C ABI (`libopenrate`), using `fiddle` from
# the standard library. No child process, no port, and no gem dependency.
#
#   require "openrate/ffi"
#
#   Openrate::Ffi.open do |engine|                 # an ENGINE. It computes.
#     engine.call("load", edges: [
#       { from: "USD", to: "ZAR", rate: 18.42, source: "my-treasury-desk" }
#     ])
#     puts engine.convert("USD", "ZAR", 100)["result"]   # => 1842.0
#   end
#
# The engine/refresher split is enforced at the ABI, not by convention:
# openrate_new builds an engine that reads no environment variable and sends no
# packet, and an engine handle REFUSES "refresh". Fetching needs a separate
# openrate_refresher_new with its own handle — see Ffi#refresher.
#
# READ sdks/ruby/README.md BEFORE YOU USE THIS. The Go runtime ends up inside
# your Ruby process and does not survive fork(), which collides with Unicorn,
# clustered Puma, Resque and Spring.
#
# Why `fiddle` and not the `ffi` gem: fiddle ships with Ruby, so the direct mode
# adds no dependency to a gem whose whole selling point is that it is thin.
# openrate's ABI has no callback at all — there is no openrate_stream — so the
# only thing needed is dlopen and a calling convention, which fiddle has.

require "fiddle"
require "json"
require "rbconfig"

module Openrate
  class Error < StandardError; end unless defined?(Error)

  # The C ABI of libopenrate (ffi/include/openrate.h) bound with fiddle.
  class Ffi
    PTR = Fiddle::TYPE_VOIDP
    U64 = Fiddle::TYPE_LONG_LONG   # fiddle has no unsigned 64-bit constant;
    VOID = Fiddle::TYPE_VOID       # handles are small counters, so this is exact.

    # Methods an ENGINE understands. A refresher method here is a clean error.
    ENGINE_METHODS = %w[convert rates meta load].freeze

    # Methods a REFRESHER understands. "refresh" and "start" open sockets.
    REFRESHER_METHODS = %w[status refresh start stop ready].freeze

    # Open an engine, yield it, and close it however the block exits. Closing an
    # engine also releases every refresher built over it, so one block is enough
    # for both.
    def self.open(config: nil, library: nil)
      engine = new(config: config, library: library)
      begin
        yield engine
      ensure
        engine.close
      end
    end

    # Build an ENGINE. config: {base: "ZAR", quiet: true}, a JSON String, or nil.
    def initialize(config: nil, library: nil)
      @library_path = library || self.class.resolve_library
      begin
        @dl = Fiddle.dlopen(@library_path)
      rescue Fiddle::DLError => e
        raise Error, "could not load #{@library_path}: #{e.message}"
      end

      bind!
      @refresher = false
      @mutex = Mutex.new

      slot = err_slot
      handle = @fn_new.call(self.class.json(config), slot)
      raise Error, "openrate_new: #{take_error(slot)}" if handle.zero?

      @handle = handle
    end

    attr_reader :library_path

    # Build a REFRESHER over this engine. It gets its own handle and its own
    # lifetime. Constructing it still opens no socket — that starts at
    # #call("refresh") or #call("start").
    def refresher(config: nil)
      assert_open!
      raise Error, "a refresher cannot own another refresher" if @refresher

      slot = err_slot
      handle = @fn_refresher_new.call(@handle, self.class.json(config), slot)
      raise Error, "openrate_refresher_new: #{take_error(slot)}" if handle.zero?

      child = clone            # shares the loaded library; only the handle differs
      child.send(:adopt_refresher_handle, handle)
      child
    end

    def refresher?
      @refresher
    end

    # The openrate version the LOADED library was built from. Static inside the
    # library — do NOT free.
    def version
      @fn_abi_version.call.to_s
    end

    # How many handles the library currently has open. Diagnostic — useful in a
    # test that wants to assert it closed what it opened.
    def open_handles
      @fn_open_handles.call
    end

    # One call against this handle. Returns the parsed JSON response.
    #
    #   engine.call("convert", from: "USD", to: "ZAR", amount: 100)
    #   engine.call("meta")
    def call(method, request = nil, **kwargs)
      request = kwargs unless kwargs.empty?
      JSON.parse(call_raw(method, request))
    end

    # The same call, returning the response JSON verbatim — the same document
    # the HTTP API publishes.
    def call_raw(method, request = nil)
      assert_open!

      # A FRESH, zeroed slot per call: openrate sets *err on failure only.
      slot = err_slot
      res = @fn_call.call(@handle, method.to_s, self.class.json(request), slot)
      raise Error, "openrate_call(#{method}): #{take_error(slot)}" if res.null?

      begin
        res.to_s   # Fiddle::Pointer#to_s copies up to the NUL into a Ruby String
      ensure
        # Everything the library returns — results AND error messages — is
        # released with openrate_free and nothing else.
        @fn_free.call(res)
      end
    end

    # Convenience for the commonest call.
    def convert(from, to, amount = 1)
      call("convert", from: from, to: to, amount: amount)
    end

    # Release this handle. Closing an ENGINE also stops and releases every
    # refresher built over it, so closing in the "wrong" order cannot leak a
    # running loop. Idempotent.
    def close
      @mutex&.synchronize do
        return if @handle.nil? || @handle.zero?

        @fn_close.call(@handle)
        @handle = 0
      end
      nil
    end

    def closed?
      @handle.nil? || @handle.zero?
    end

    # ---------------------------------------------------------------- internals

    private

    def adopt_refresher_handle(handle)
      @handle = handle
      @refresher = true
      @mutex = Mutex.new
      self
    end

    def bind!
      # need_gvl: is left at its default (false), so fiddle routes the call
      # through rb_thread_call_without_gvl and other Ruby threads keep running
      # during a refresh that is waiting on a central bank. Nothing in
      # openrate's ABI calls back into Ruby — there is no openrate_stream — so
      # there is no closure to worry about here at all.
      @fn_abi_version = Fiddle::Function.new(@dl["openrate_abi_version"], [], PTR,
                                             name: "openrate_abi_version")
      @fn_new = Fiddle::Function.new(@dl["openrate_new"], [PTR, PTR], U64, name: "openrate_new")
      @fn_refresher_new = Fiddle::Function.new(@dl["openrate_refresher_new"], [U64, PTR, PTR], U64,
                                               name: "openrate_refresher_new")
      @fn_call = Fiddle::Function.new(@dl["openrate_call"], [U64, PTR, PTR, PTR], PTR,
                                      name: "openrate_call")
      @fn_close = Fiddle::Function.new(@dl["openrate_close"], [U64], VOID, name: "openrate_close")
      @fn_free = Fiddle::Function.new(@dl["openrate_free"], [PTR], VOID, name: "openrate_free")
      @fn_open_handles = Fiddle::Function.new(@dl["openrate_open_handles"], [], U64,
                                              name: "openrate_open_handles")
    rescue Fiddle::DLError => e
      raise Error, "#{@library_path} is missing a symbol the ABI promises: #{e.message}"
    end

    def assert_open!
      raise Error, "this Openrate::Ffi handle is closed" if closed?
    end

    # A zeroed `char*` slot for the trailing `char** err` argument.
    def err_slot
      slot = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP, Fiddle::RUBY_FREE)
      slot[0, Fiddle::SIZEOF_VOIDP] = "\x00".b * Fiddle::SIZEOF_VOIDP
      slot
    end

    # Read the message out of an err slot and free it with openrate_free — an
    # error must not be a leak.
    def take_error(slot)
      ptr = slot.ptr
      return "(no message)" if ptr.null?

      message = ptr.to_s
      @fn_free.call(ptr)
      message
    end

    class << self
      def json(value)
        case value
        when nil, String then value
        else JSON.generate(value)
        end
      end

      # 1. OPENRATE_LIBRARY
      # 2. lib/ inside this gem (where a platform gem would put it)
      # 3. dist/ffi/ in a checkout — note the naming is
      #    libopenrate-<goos>-<goarch>.<ext>, not a per-target directory
      # 4. the bare soname, letting the dynamic loader search its own paths
      def resolve_library
        env = ENV["OPENRATE_LIBRARY"]
        return env if env && !env.empty?

        ext = library_extension
        gem_root = File.expand_path("../..", __dir__) # …/sdks/ruby
        repo = File.expand_path("../..", gem_root)    # …/repo root

        candidates = [
          File.join(gem_root, "lib", "libopenrate.#{ext}"),
          File.join(repo, "dist", "ffi", "libopenrate-#{goos}-#{goarch}.#{ext}")
        ]
        candidates.find { |path| File.file?(path) } ||
          (goos == "windows" ? "openrate.dll" : "libopenrate.#{ext}")
      end

      def library_extension
        case RbConfig::CONFIG["host_os"]
        when /mswin|mingw|cygwin/ then "dll"
        when /darwin/ then "dylib"
        else "so"
        end
      end

      def goos
        case RbConfig::CONFIG["host_os"]
        when /mswin|mingw|cygwin/ then "windows"
        when /darwin/ then "darwin"
        when /freebsd/ then "freebsd"
        else "linux"
        end
      end

      def goarch
        case RbConfig::CONFIG["host_cpu"]
        when /arm64|aarch64/ then "arm64"
        when /x86_64|amd64/ then "amd64"
        else RbConfig::CONFIG["host_cpu"]
        end
      end
    end
  end
end

File: sdks/php/src/Ffi.php
<?php

declare(strict_types=1);

namespace Openrate;

/**
 * openrate in-process, through the C ABI (`libopenrate`), using PHP's bundled
 * FFI extension. No child process, no port.
 *
 *   $engine = new \Openrate\Ffi();               // an ENGINE. It computes.
 *   try {
 *       $engine->call('load', ['edges' => [...]]);          // rates you obtained
 *       $r = $engine->call('convert', ['from' => 'USD', 'to' => 'ZAR', 'amount' => 100]);
 *   } finally {
 *       $engine->close();
 *   }
 *
 * The engine/refresher split is enforced at the ABI, not by convention:
 * openrate_new() builds an engine that reads no environment variable and sends
 * no packet, and an engine handle REFUSES "refresh". Fetching needs a separate
 * openrate_refresher_new() with its own handle — see Ffi::refresher().
 *
 * READ sdks/php/README.md BEFORE YOU USE THIS. The Go runtime ends up inside
 * your PHP process and it is not fork-safe, which collides head-on with
 * php-fpm. For most PHP deployments \Openrate\Openrate — the managed sidecar —
 * is the right answer.
 */
final class Ffi
{
    /**
     * The C ABI, transcribed from ffi/include/openrate.h. PHP's FFI parser does
     * not run the C preprocessor, so the header itself cannot be handed to
     * FFI::cdef. Note what is NOT here: there is no openrate_stream, because
     * openrate answers from a snapshot and has nothing to stream.
     */
    private const CDEF = <<<'C'
        const char* openrate_abi_version(void);
        uint64_t openrate_new(const char* config_json, char** err);
        uint64_t openrate_refresher_new(uint64_t engine, const char* config_json, char** err);
        char* openrate_call(uint64_t h, const char* method, const char* request_json, char** err);
        void openrate_close(uint64_t h);
        void openrate_free(char* p);
        uint64_t openrate_open_handles(void);
        C;

    /** Methods an ENGINE understands. A refresher method here is a clean error. */
    public const ENGINE_METHODS = ['convert', 'rates', 'meta', 'load'];

    /** Methods a REFRESHER understands. "refresh" and "start" open sockets. */
    public const REFRESHER_METHODS = ['status', 'refresh', 'start', 'stop', 'ready'];

    /** @var \FFI */
    private $ffi;

    /** @var int */
    private $handle = 0;

    /** @var string */
    private $libraryPath;

    /** @var bool true for a refresher handle, false for an engine */
    private $isRefresher;

    /** @var self|null the engine a refresher was built over, for close ordering */
    private $engine;

    /**
     * Build an ENGINE.
     *
     * @param array{base?:string,quiet?:bool}|string|null $config
     */
    public function __construct($config = null, ?string $libraryPath = null)
    {
        if (!\extension_loaded('FFI')) {
            throw new OpenrateException(
                'the FFI extension is not loaded. It ships with PHP >= 7.4 but is ' .
                'often disabled: enable extension=ffi in php.ini. See sdks/php/README.md.'
            );
        }

        $this->libraryPath = $libraryPath ?? self::resolveLibrary();
        $this->isRefresher = false;
        $this->engine = null;

        try {
            $this->ffi = \FFI::cdef(self::CDEF, $this->libraryPath);
        } catch (\FFI\Exception $e) {
            throw new OpenrateException(
                'FFI::cdef failed for ' . $this->libraryPath . ': ' . $e->getMessage() .
                (\strpos($e->getMessage(), 'ffi.enable') !== false
                    ? "\nffi.enable is 'preload' by default, which forbids FFI::cdef in " .
                      'any SAPI except CLI. See sdks/php/README.md — the sidecar ' .
                      '(\\Openrate\\Openrate) needs no php.ini change at all.'
                    : ''),
                0,
                $e
            );
        }

        $err = $this->ffi->new('char*');
        $handle = $this->ffi->openrate_new(self::json($config), \FFI::addr($err));
        if ($handle === 0) {
            throw new OpenrateException('openrate_new: ' . $this->takeError($err));
        }
        $this->handle = (int) $handle;
    }

    /**
     * Run $fn with an open engine and close it afterwards, whatever happens.
     * Closing an engine also closes every refresher built over it, so this one
     * `finally` is enough for both.
     *
     * @param array{base?:string,quiet?:bool}|string|null $config
     * @return mixed
     */
    public static function with($config, callable $fn, ?string $libraryPath = null)
    {
        $engine = new self($config, $libraryPath);
        try {
            return $fn($engine);
        } finally {
            $engine->close();
        }
    }

    /**
     * Build a REFRESHER over this engine. It gets its own handle and its own
     * lifetime. Constructing it still opens no socket — that starts at
     * `$refresher->call('refresh')` or `->call('start')`.
     *
     * @param array{sources?:string,interval_ms?:int,fetch_timeout_ms?:int,quiet?:bool}|string|null $config
     */
    public function refresher($config = null): self
    {
        $this->assertOpen();
        if ($this->isRefresher) {
            throw new OpenrateException('a refresher cannot own another refresher');
        }

        $err = $this->ffi->new('char*');
        $handle = $this->ffi->openrate_refresher_new($this->handle, self::json($config), \FFI::addr($err));
        if ($handle === 0) {
            throw new OpenrateException('openrate_refresher_new: ' . $this->takeError($err));
        }

        // A clone shares the loaded FFI object; only the handle differs.
        $refresher = clone $this;
        $refresher->handle = (int) $handle;
        $refresher->isRefresher = true;
        $refresher->engine = $this;

        return $refresher;
    }

    /** The openrate version the LOADED library was built from. Never free it. */
    public function version(): string
    {
        return (string) $this->ffi->openrate_abi_version();
    }

    /** How many handles the library currently has open. Diagnostic. */
    public function openHandles(): int
    {
        return (int) $this->ffi->openrate_open_handles();
    }

    /** The path this instance actually loaded. */
    public function libraryPath(): string
    {
        return $this->libraryPath;
    }

    public function isRefresher(): bool
    {
        return $this->isRefresher;
    }

    /**
     * One call against this handle.
     *
     * @param array<string,mixed>|string|null $request
     * @return array<string,mixed> the decoded JSON response
     */
    public function call(string $method, $request = null): array
    {
        $decoded = json_decode($this->callRaw($method, $request), true);
        if (!\is_array($decoded)) {
            throw new OpenrateException("openrate returned a non-object response to '{$method}'");
        }

        return $decoded;
    }

    /**
     * The same call, returning the response JSON verbatim — the same document
     * the HTTP API publishes.
     *
     * @param array<string,mixed>|string|null $request
     */
    public function callRaw(string $method, $request = null): string
    {
        $this->assertOpen();

        // A FRESH, zeroed slot per call: openrate sets *err on failure only.
        $err = $this->ffi->new('char*');
        $res = $this->ffi->openrate_call($this->handle, $method, self::json($request), \FFI::addr($err));

        // A NULL char* return arrives in PHP as null, not as a null CData.
        if ($res === null) {
            throw new OpenrateException("openrate_call({$method}): " . $this->takeError($err));
        }

        try {
            return \FFI::string($res);
        } finally {
            $this->ffi->openrate_free($res);
        }
    }

    /** Convenience for the commonest call. */
    public function convert(string $from, string $to, float $amount = 1.0): array
    {
        return $this->call('convert', ['from' => $from, 'to' => $to, 'amount' => $amount]);
    }

    /**
     * Release this handle. Closing an ENGINE also stops and releases every
     * refresher built over it, so closing in the "wrong" order cannot leak a
     * running loop. Idempotent.
     */
    public function close(): void
    {
        if ($this->handle !== 0) {
            $this->ffi->openrate_close($this->handle);
            $this->handle = 0;
        }
    }

    /** A safety net. close() in a finally block is the mechanism. */
    public function __destruct()
    {
        $this->close();
    }

    // ---------------------------------------------------------------- internals

    private function assertOpen(): void
    {
        if ($this->handle === 0) {
            throw new OpenrateException('this Openrate\Ffi handle is closed');
        }
    }

    /**
     * Read the message out of an err slot and free it with openrate_free — an
     * error must not be a leak.
     *
     * @param \FFI\CData $err
     */
    private function takeError($err): string
    {
        if (\FFI::isNull($err)) {
            return '(no message)';
        }
        $message = \FFI::string($err);
        $this->ffi->openrate_free($err);

        return $message;
    }

    /** @param array<string,mixed>|string|null $value */
    private static function json($value): ?string
    {
        if ($value === null || \is_string($value)) {
            return $value;
        }
        $json = json_encode($value);
        if ($json === false) {
            throw new OpenrateException('could not encode the request as JSON: ' . json_last_error_msg());
        }

        return $json;
    }

    /**
     * Find libopenrate:
     *   1. OPENRATE_LIBRARY
     *   2. lib/ inside this package
     *   3. dist/ffi/ in a checkout of the openrate repo — note the naming is
     *      libopenrate-<goos>-<goarch>.<ext>, not a per-target directory
     *   4. the bare soname, letting the dynamic loader search its own paths
     */
    private static function resolveLibrary(): string
    {
        $env = getenv('OPENRATE_LIBRARY');
        if ($env !== false && $env !== '') {
            return $env;
        }

        $ext = self::libraryExtension();
        $pkg = \dirname(__DIR__);   // .../sdks/php
        $repo = \dirname($pkg, 2);  // repo root

        $candidates = [
            $pkg . '/lib/libopenrate.' . $ext,
            $repo . '/dist/ffi/libopenrate-' . self::goos() . '-' . self::goarch() . '.' . $ext,
        ];
        foreach ($candidates as $candidate) {
            if (is_file($candidate)) {
                return $candidate;
            }
        }

        return \PHP_OS_FAMILY === 'Windows' ? 'openrate.dll' : 'libopenrate.' . $ext;
    }

    private static function libraryExtension(): string
    {
        switch (\PHP_OS_FAMILY) {
            case 'Windows':
                return 'dll';
            case 'Darwin':
                return 'dylib';
            default:
                return 'so';
        }
    }

    private static function goos(): string
    {
        switch (\PHP_OS_FAMILY) {
            case 'Windows':
                return 'windows';
            case 'Darwin':
                return 'darwin';
            case 'BSD':
                return 'freebsd';
            default:
                return 'linux';
        }
    }

    private static function goarch(): string
    {
        $machine = strtolower(php_uname('m'));
        if ($machine === 'arm64' || $machine === 'aarch64') {
            return 'arm64';
        }
        if ($machine === 'x86_64' || $machine === 'amd64') {
            return 'amd64';
        }

        return $machine;
    }
}
